airbyte-integrations/connectors/destination-s3-v2/src/main/kotlin/S3V2Writer.kt

```diff
@@ -63,9 +87,41 @@ class S3V2Writer(
                 pathFactory
                     .getPathToFile(stream, stagedObject.partNumber, isStaging = false)
                     .toString()
+            log.info { "Moving staged object from ${stagedObject.s3Object.key} to $finalKey" }
             val newObject = s3Client.move(stagedObject.s3Object, finalKey)
+
+            val state = destinationStateManager.getState(stream)
+            state.removeObject(stream.generationId, stagedObject.s3Object.key)
+            state.addObject(stream.generationId, newObject.key, stagedObject.partNumber)
```

Comment:
is it possible for state to be persisted while this function is running? (i.e. do we need to have a mutex somewhere)

currently it looks like we're only persistState from `close()`, so we're probably ok? But do we want to eventually persist state during a sync?

Reply:
Right now I'm assuming one persistence at the end. Ultimately we'll have to enforce that these operations are serialized (push to a queue, process in order) with the ability to block when you want to await confirmation about persistence.

Right now I'll just put a lock on the table updates.